### PR TITLE
Build the chef schema and endpoint

### PR DESCRIPTION
**Pull request by harethriyadh**

### CODE REVIEW

File: src/routes/dish.js [hunk 1]
```diff
@@ -7,5 +7,6 @@ router.use(express.json());
 
 router.post('/', dishController.createDish);
 router.get('/', dishController.getAllDishes);
+router.post('/getDish', dishController.getDish);
```
Comment (motaz99) [inline src/routes/dish.js]: We don't need this, Tesnim is already built it. The only thing that related to your task is to create a dish to become related to a chef, and also get all dishes that are related to the chef. And Tesnim is already built those endpoints and controllers. You just need to connect them to your chef.

File: src/routes/chef.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const express = require('express');
+const chef = require('../controllers/chefController');
+const dish = require('../controllers/dishController');
+
+const router = express.Router();
+
+router.post('/', chef.createRestaurant);
+router.get('/me', chef.getRestaurantInfo);
+router.put('/me', chef.editRestaurant);
```
Comment (motaz99) [inline src/routes/chef.js]: Those name need to changed to fit with the paths.

They should be:
router.post('/', chef.createChef);
router.get('/me', chef.getChefInfo);
router.put('/me', chef.editChefInfo);

File: src/models/chef.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+const mongoose = require('mongoose');
+
+const chefSchema = new mongoose.Schema({
+  userId: {
+    type: String,
+    required: true,
+  },
```
Comment (motaz99) [inline src/models/chef.js]: This need to be commented for now, cuz you shouldn't care about the user for now.

File: src/models/chef.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  ContactNumber: {
+    type: String,
+    required: true,
+  },
+  Description: {
+    type: String,
+    required: true,
+  },
```
Comment (motaz99) [inline src/models/chef.js]: You shouldn't start with capital letters with your schema properties.

We need to change them.

File: src/models/dish.js [hunk 5]
```diff
@@ -2,6 +2,10 @@ const mongoose = require('mongoose');
 
 const dishSchema = new mongoose.Schema({
   // chef:{ type: Schema.Types.ObjectId, ref: 'Chef' },
+  userId: {
+    type: String,
+    required: true,
+  },
```
Comment (motaz99) [inline src/models/dish.js]: Why did you add userId?
We don't need it here, we will not put the userId inside the dish schema.
Even though this is not the right way user feature of objectId of mongo. 
This is the right way to do it 
`chef: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Chef',
  },`

And also if you can notice that Tesnim is already added this for you, you just need to uncomment it from the chef schema.

File: src/controllers/chefController.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+exports.createRestaurant = async (req, res) => {
+  const {
+    Restaurant,
+    Location,
+    OpeningHours,
+    ClosingHours,
+    ContactNumber,
+    Description,
+  } = req.body;
```
Comment (motaz99) [inline src/controllers/chefController.js]: Why are these variable names start with capital letters?
We need to change them.

File: src/controllers/chefController.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    if (
+      !Restaurant &&
+      !Location &&
+      !OpeningHours &&
+      !ClosingHours &&
+      !ContactNumber &&
+      !Description
+    ) {
+      throw new Error('Incomplete information...');
+    }
```
Comment (motaz99) [inline src/controllers/chefController.js]: We don't need this, this is already handled by the chef schema you wrote, you already put a required for all of them, so if there is any incomplete information, mongoose will through and error, and you can even specify that error inside the schema. 
You can take a look on this [blog](https://levelup.gitconnected.com/handling-errors-in-mongoose-express-for-display-in-react-d966287f573b)

File: src/controllers/chefController.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
+const Chef = require('../models/chef');
+
+exports.createRestaurant = async (req, res) => {
```
Comment (motaz99) [inline src/controllers/chefController.js]: This function name should `createChef`

File: src/controllers/chefController.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      !Location &&
+      !OpeningHours &&
+      !ClosingHours &&
+      !ContactNumber &&
+      !Description
+    ) {
+      throw new Error('Incomplete information...');
+    }
+
+    const checkRestaurant = await Chef.findOne({ Restaurant });
```
Comment (motaz99) [inline src/controllers/chefController.js]: This doesn't make sense, the  varible name is `checkRestaurant` and the schema name is `Chef`

`checkRestaurant` needs to be changed to `chef`

File: src/controllers/chefController.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    if (checkRestaurant) {
+      throw new Error('Restaurant already exists.');
+    }
+
+    const newRestaurant = new Chef({
```
Comment (motaz99) [inline src/controllers/chefController.js]: This one as well. Every restauran need to be changed to a chef.

File: src/controllers/chefController.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      userId: req.session.userId,
+      Restaurant,
+      Location,
+      OpeningHours,
+      ClosingHours,
+      ContactNumber,
+      Description,
+    });
+
+    await newRestaurant.save();
```
Comment (motaz99) [inline src/controllers/chefController.js]: Again, Don't start with capital letters.

Also one more thing, we are not using session, we are using JWT and cookie. Even you wrote a file named JWT, so we either use JWT and cookie, or we use the session. And I think that we already went with the JWT and Cookie.

File: src/controllers/chefController.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+      throw new Error(
+        'Restaurant does not exist. You need to create your restaurant.'
+      );
+    }
+
+    res.status(200).json({ restaurant });
+  } catch (error) {
+    res.status(500).json({ error: 'Server error' });
+  }
+};
```
Comment (motaz99) [inline src/controllers/chefController.js]: Replace the restaurant with a chef.

File: src/controllers/chefController.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    if (
+      !Restaurant &&
+      !Location &&
+      !OpeningHours &&
+      !ClosingHours &&
+      !ContactNumber &&
+      !Description
+    ) {
+      throw new Error('Incomplete information...');
+    }
```
Comment (motaz99) [inline src/controllers/chefController.js]: Mongoose can handle this, we don't need this condition. I mentioned this above.

File: src/controllers/chefController.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+exports.editRestaurant = async (req, res) => {
+  const id = req.session.userId;
+  const {
+    Restaurant,
+    Location,
+    OpeningHours,
+    ClosingHours,
+    ContactNumber,
+    Description,
+  } = req.body;
```
Comment (motaz99) [inline src/controllers/chefController.js]: don't start with capital letters.

File: src/controllers/chefController.js [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+    res.status(200).json({ restaurant });
+  } catch (error) {
+    res.status(500).json({ error: 'Server error' });
+  }
+};
+
+// Edit the restaurant info
+exports.editRestaurant = async (req, res) => {
+  const id = req.session.userId;
```
Comment (motaz99) [inline src/controllers/chefController.js]: We don't use session. we use JWT and cookie.

This is also is out of your task range. You don't need to care about the session or even the JWT at this point.

File: src/controllers/chefController.js [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    if (!checkRestaurant) {
+      throw new Error('Restaurant does not exist');
+    }
+
+    checkRestaurant.Restaurant = Restaurant;
+    checkRestaurant.Location = Location;
+    checkRestaurant.OpeningHours = OpeningHours;
+    checkRestaurant.ClosingHours = ClosingHours;
+    checkRestaurant.ContactNumber = ContactNumber;
+    checkRestaurant.Description = Description;
```
Comment (motaz99) [inline src/controllers/chefController.js]: What will happen if one of the variables you extract from the is `undefined`? are you going to store `undefined` value to the DB?

File: src/controllers/dishController.js [hunk 17]
```diff
@@ -32,9 +36,24 @@ exports.getAllDishes = async (req, res) => {
...
+    const checkDish = await Dish.findOne({ userId: id });
+    if (!checkDish) {
+      throw new Error('Dish does not exist...');
+    }
+
+    res.status(200).json({ dish: checkDish });
+  } catch (error) {
+    res.status(500).json({ error: error.message });
+  }
+};
```
Comment (motaz99) [inline src/controllers/dishController.js]: This is out of your task range.
Tesnim is already built this in a another PR

File: src/controllers/dishController.js [hunk 18]
```diff
@@ -18,10 +23,9 @@ exports.createDish = async (req, res) => {
     const savedDish = await newDish.save();
 
     // Send the saved dish as a response
-    res.status(201).json(savedDish);
+    res.status(201).json({ savedDish });
   } catch (error) {
-    // Handle server error
-    res.status(500).json({ error: 'Internal server error' });
+    res.status(500).json(error.message);
   }
```
Comment (motaz99) [inline src/controllers/dishController.js]: Inside the createDish, all you need to do is to put the Id of the chef inside the params object that inside the req object.

So for example when you do a request from postman to create a dish, what you need to do is just to put the id of the chef that you want the dish to be related to manually, so when the dish get create it will has the id of a chef that you add his Id manually.

File: src/controllers/authController.js
Comment (motaz99) [inline src/controllers/authController.js:1]: We are not going to use a session, and you don't need to worry about his for now, all you have to do is to create a chef, and if create dishes that related to that chef, and you are going to add the IDs manually, That is it for now. Because I am working on the login possess, and if you add a session and I already added JWT and cookie our code will be in a chaous.

File: src/app.js
Comment (motaz99) [inline src/app.js:1]: We don't need to add a session.

File: package.json
Comment (motaz99) [inline package.json:1]: We don't need express-session

File: src/jwt.js
Comment (motaz99) [inline src/jwt.js:1]: This is not your task, we already talked about this, don't do anything is not related to your task. We need to get this file back.

**Review by motaz99 — CHANGES_REQUESTED** [review]

File: src/controllers/authController.js
Comment (harethriyadh) [reply src/controllers/authController.js:1]: I need to test it.

File: src/app.js
Comment (harethriyadh) [reply src/app.js:1]: I need it for testing...

File: package.json
Comment (harethriyadh) [reply package.json:1]: I need it for testing....

File: src/jwt.js
Comment (harethriyadh) [reply src/jwt.js:1]: you already deleted it from your task also!!!

File: src/routes/dish.js (hunk 1 above)
Comment (harethriyadh) [reply src/routes/dish.js]: did it!!

File: src/models/chef.js
Comment (motaz99) [inline src/models/chef.js:1]: As a general feedback for for the tow changes you did, inside this commit as the following:

1. Your commit message says `change chef routes names` 
2.but actually you changed the model and commented the userId inside the dish and the chef schemas
3. So always try to make your commit fit with your things you do in the code, it will helps the reviewer understand what are you doing in this commit.

Good work!

File: src/routes/chef.js
Comment (motaz99) [inline src/routes/chef.js:1]: Oh, now I see, I think you flipped the commit messages by mistake.

File: src/controllers/chefController.js [hunk 19]
```diff
@@ -2,39 +2,39 @@ const Chef = require('../models/chef');
...
-      !Description
+      !restaurant &&
+      !location &&
+      !openingHours &&
+      !closingHours &&
+      !contactNumber &&
+      !description
     ) {
       throw new Error('Incomplete information...');
     }
```
Comment (motaz99) [inline src/controllers/chefController.js]: Great work on changing the variable names! 
But as we said we don't need these conditions, because mongoose is going to handle it. So you you can delete it, we don't need it.

File: src/controllers/chefController.js [hunk 20]
```diff
@@ -2,39 +2,39 @@ const Chef = require('../models/chef');
...
 
-    const checkRestaurant = await Chef.findOne({ Restaurant });
```
Comment (motaz99) [inline src/controllers/chefController.js]: You didn't change this `checkChef`, Everything that refer to a the chef schema we wrote, and you named restaurant need to be changed to chef.

File: src/controllers/chefController.js [hunk 21]
```diff
@@ -1,6 +1,8 @@
 const Chef = require('../models/chef');
 
-exports.createRestaurant = async (req, res) => {
+exports.createChef = async (req, res) => {
+  // we need to get an id as an authorized user
+  // const id = req.
```
Comment (motaz99) [inline src/controllers/chefController.js:5]: I like this comment, good one.

**Review by motaz99 — APPROVED** [review]
LGTM! Wohoooo we made it.
You still have some conflicts, would like me to help you solving them?

**Review by ozlemkeles — APPROVED** [review]
it looks better now, good work, thank you

File: src/controllers/chefController.js (hunk 20 above)
Comment (harethriyadh) [reply src/controllers/chefController.js]: It's the schema token variable that is taken from the req.body

**Comment by harethriyadh** [comment]
Closing #27

**Comment by harethriyadh** [comment]
Closing #27
